### DISCUSSION TITLE
How to reload without cache?

### DISCUSSION THREAD

**Discussion by frankdekker** [MicrosoftEdge/WebView2Feedback#3923] [Q&A]
Currently the CoreWebView2.Reload method reloads the currently active page preserving the cache state. Is there a way to programmatically reload the page without keeping the cache?

**Comment by hunkydoryrepair** [comment]
yes, must use CTRL-SHIFT-R to clear without cache, but I can't figure out how to do it in code, either, other than use javascript.

**Comment by frankdekker** [comment]
Could you give an example how you would reload it via javascript? My current experimentation is to send CTRL+SHIFT+R to the WebView2 window handle, not quite successful so far.

**Comment by hunkydoryrepair** [comment]
Well, it used to work by doing this:

```
ctlWebBrowser.ExecuteScriptAsync("window.location.reload(true)");
```

I find that is no longer sufficient. It reloads the main document but not javascript or css files.  The `true` parameter was never official so chromium must have dropped support for it.

**Comment by torgeirf** [comment]
We have now had 2 occasions where this feature would be beneficial. First occasion was an unintentional permanent redirect response from the server, and the second case was a corrupt js file. Both of these issues required either:

1) Deleting the user data folder
or
2) Opening dev tools (guiding the user via phone) and make him disable cache and reload again

Please consider a method to "hard reload" the page programatically.

**Comment by mbodm** [comment]
Something happened along those lines (since Jan 2022)?

I also have a hard time here, to figure out how i can completely "re-open" a site, in WebView2. Since we have a system that generates CSS and JS, along with HTML and we have to "re-render" this, by refreshing the whole site in WebView2.

The things i tried:
- `webView2.Source = new Uri("about:blank");` followed by `webView2.Source = new Uri("http://mywebsite.com");`
- Above mentioned JS call: `await webView2.ExecuteScriptAsync("window.location.reload(true)");`
Info: This does not work in general, i.e. in the console window in Edge browser.
- `webView2.Reload();`

None of that helped. So i have no clue at the moment, how a WebView2 could do a completely fresh load of a website.

PS: The only thing that works, is to open the browser tools in WebView2 and press CTRL+F5.

**Comment by mbodm** [comment]
**Additional note**: I found 1 another "trick" to force a full reload of the site (including CSS and JS), by invalidating the cache:

```
await webView2.CoreWebView2.Profile.ClearBrowsingDataAsync();
webView2.Reload();
```

The only other option was sending a CTRL+F5 KeyPress event to the WebView2 container.

Maybe this helps someone coming here for the same questions.

**Reply by sln162** [reply]
Isn't this clearing the cache for all UDF locations?  Isn't it the same as re-designating new UDF addresses?  If there are many caches, deleting all cache files may take time.I think it's a loss.

**Reply by mbodm** [reply]
Good question @sln162 

Since in my use case i don’t care much about the UDF, i don’t know, tbh.

**Comment by sln162** [comment]
Compromise solution:  
 ```
  var uriStr = webView21.Source.AbsoluteUri;
  
  if (string.IsNullOrEmpty(webView21.Source.Query))
  {
      uriStr += "?t=" + DateTime.Now.Ticks.ToString();
  }
  else if (webView21.Source.Query.Contains("t="))
  {
      uriStr = Regex.Replace(uriStr, "t=([^&]*)", "t=" + DateTime.Now.Ticks.ToString());
  }
  else
  {
      uriStr += "&t=" + DateTime.Now.Ticks.ToString();
  }
  
  webView21.CoreWebView2.Navigate(uriStr);
```

**Reply by Sandu119** [reply]
Just an idea (never tried it): inject the new css into the html.
